### DISCUSSION TITLE
Are threads in a wavefront expected to always be in lockstep, regardless of masked lanes?

### DISCUSSION THREAD

**Discussion by ashesh2512** [ROCm/ROCm#4033] [Q&A]
If I have a conditional statement that masks certain lanes in a wavefront, is there any possibility for the masked/no-op lanes to race ahead of the rest of the wavefront, or will they always wave for the conditional code to finish before moving forward?

**Comment by b-sumner** [comment]
Hi @ashesh2512 it's generally not a good idea to count on the scheduling of other threads in general.  Its also not a good idea to count on a ballot producing what you think it will in any given language constructs.    Are you seeing an unexpected result, or doing pre-development research?

**Reply by ashesh2512** [reply]
@b-sumner I have a conditional statement that is valid only for certain lanes. With CUDA, I need to add a warp sync after the conditional statement, else the solution is wrong. However, with ROCm, I observed not needing any synchronization after the conditional statement, and hence was wondering if a wavefront is always in lockstep.

**Comment by b-sumner** [comment]
@ashesh2512 when a wavefront executes an instruction it executes it on behalf of all of the threads/lanes in that wavefront that are active at the time.  But you cannot know what lanes will be active (i.e. the nonzero bits in a ballot(1)) by looking at source code.
